docs/design/design.md

```diff
@@ -0,0 +1,261 @@
+## Implementation design for kubeadm
+
+`kubeadm init` and `kubeadm join` together provides a nice user experience for creating a best-practice but bare Kubernetes cluster from scratch.
+However, it might not be obvious _how_ kubeadm does that.
+
+This document strives to explain what types of phases of work happen under the hood.
+Also included is ComponentConfiguration API types for talking to kubeadm programmatically.
```

Comment:
Are you relying on the componentconfig API group? Be careful with this, that group will be deleted soon (as soon as every component's config has been moved out of it).